pallets/asset-index/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+        // Origin that is allowed to administer the index
+        type AdminOrigin: EnsureOrigin<Self::Origin>;
+        // Currency implementation to use as the index token
+        type IndexToken: LockableCurrency<Self::AccountId>;
+        // Period after the minting of the index token for which 100% is locked up.
+        // Only applies to users contributing assets directly to index
+        type LockupPeriod: Get<Self::BlockNumber>;
+        // The minimum amount of the index token that can be redeemed for the underlying asset in the index
+        type MinimumRedemption: Get<BalanceFor<Self>>;
+        // Minimum amount of time between redeeming index tokens
+        // and being able to withdraw the awarded assets
+        type WithdrawalPeriod: Get<Self::BlockNumber>;
+        // The maximum amount of DOT that can exist in the index
+        type DOTContributionLimit: Get<BalanceFor<Self>>;
+        // Type used to identify assets
+        type AssetId: Parameter + Encode + Decode;
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    /// (AssetId) -> IndexAssetData
+    pub type Holdings<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AssetId, IndexAssetData<BalanceFor<T>>, OptionQuery>;
+
+    #[pallet::storage]
+    /// (AccountId) -> Balance. Tracks how much each LP has contributed
+    pub type Depositors<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, BalanceFor<T>, OptionQuery>;
+
+    #[pallet::storage]
+    ///  (AccountId) -> Vec<PendingRedemption>
+    pub type PendingWithdrawals<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        Vec<PendingRedemption<T::AssetId, BalanceFor<T>, BlockNumberFor<T>>>,
+        OptionQuery,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        // A new asset was added to the index and some index token paid out
+        // \[AssetIndex, AssetUnits, IndexTokenRecipient, IndexTokenPayout\]
+        AssetAdded(T::AssetId, BalanceFor<T>, AccountIdFor<T>, BalanceFor<T>),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        // Thrown if adding units to an asset holding causes it numerical type to overflow
+        AssetUnitsOverflow,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(10_000)] // TODO: Set weights
+        /// Callable by an admin to add new assets to the index and mint some IndexToken
+        /// Caller balance is updated to allocate the correct amount of the IndexToken
+        /// Creates IndexAssetData if it doesn’t exist, otherwise adds to list of deposits
+        pub fn add_asset(
```

Comment:
I suppose this clippy warning will go away when we add more extrinsics.

Reply:
Not much we can do for this one I think since it comes from macro generated code